### PR TITLE
Docker local changes

### LINKED ISSUE #703: Newest start_docker.sh file doesn't allow for local changes to code
**Describe the bug**
After pulling the current codebase on main, I found that local changes to any of the scripts were not being updated in Docker (same problem I was having before, which was caused by the wrong docker tag in start_docker.sh). I rolled back a few commits and figured out the issue. 

In this commit https://github.com/angelolab/ark-analysis/commit/f3044e78d5b683dd7a06dbe08dc130baae7aecdb, the following line was removed from start_docker.sh (in 2 places):
`-v "$PWD/ark:/usr/local/lib/python3.7/site-packages/ark" \`

If I understand correctly, that mounts the codebase to docker so any local code changes are correctly propagated. I added that line back in (but changed it to python3.8) to the version of start_docker.sh currently on main, and it seemed to have solved the issue. Relevant code block looks like this:

```
if [ ! -z "$external" ]
  then
    docker run -it \
      -p $PORT:$PORT \
      -e JUPYTER_PORT=$PORT \
      -e JUPYTER_DIR=$JUPYTER_DIR \
      -v "$PWD/ark:/usr/local/lib/python3.8/site-packages/ark" \
      -v "$PWD/scripts:/scripts" \
      -v "$PWD/data:/data" \
      -v "$external:/data/external" \
      -v "$PWD/ark/phenotyping/create_pixel_som.R:/create_pixel_som.R" \
      -v "$PWD/ark/phenotyping/run_pixel_som.R:/run_pixel_som.R" \
      -v "$PWD/ark/phenotyping/pixel_consensus_cluster.R:/pixel_consensus_cluster.R" \
      -v "$PWD/ark/phenotyping/create_cell_som.R:/create_cell_som.R" \
      -v "$PWD/ark/phenotyping/run_cell_som.R:/run_cell_som.R" \
      -v "$PWD/ark/phenotyping/cell_consensus_cluster.R:/cell_consensus_cluster.R" \
      -v "$PWD/.toks:/home/.toks" \
      angelolab/ark-analysis:v0.4.1
  else
    docker run -it \
      -p $PORT:$PORT \
      -e JUPYTER_PORT=$PORT \
      -e JUPYTER_DIR=$JUPYTER_DIR \
      -v "$PWD/ark:/usr/local/lib/python3.8/site-packages/ark" \
      -v "$PWD/scripts:/scripts" \
      -v "$PWD/data:/data" \
      -v "$PWD/ark/phenotyping/create_pixel_som.R:/create_pixel_som.R" \
      -v "$PWD/ark/phenotyping/run_pixel_som.R:/run_pixel_som.R" \
      -v "$PWD/ark/phenotyping/pixel_consensus_cluster.R:/pixel_consensus_cluster.R" \
      -v "$PWD/ark/phenotyping/create_cell_som.R:/create_cell_som.R" \
      -v "$PWD/ark/phenotyping/run_cell_som.R:/run_cell_som.R" \
      -v "$PWD/ark/phenotyping/cell_consensus_cluster.R:/cell_consensus_cluster.R" \
      angelolab/ark-analysis:v0.4.1
fi
```
After doing this, any local changes I made to main were then properly propagated to Docker.

**Expected behavior**
I'm not sure why those lines were initially removed, but whatever the reason or if there is a better solution than the one I used, I think it is important for users to be able to make local edits to the code and have them reflected in Docker.

**To Reproduce**
Pull version of code on main and change any of the scripts. They are not reflected inside of Docker.

### PR DESCRIPTION
**Pull request by ackagel**
## **Purpose**

Allows users to update+reinstall ark codebase within docker with `-u` flag. closes #703 

## **Implementation**

Adds the `ARK_UPDATE` environment variable, which is set to 0 or 1 depending on if the `-u` or `--update` flag is set.  If true, the docker will reinstall ark-analysis before running the jupyter lab command.  This is done by replacing the `CMD jupyter lab ...` line in `Dockerfile` with a call to the new `start_jupyter.sh` script.

## **Remaining issues**

I think I can neatly fit #710 in here too, but this is ready for testing in the mean time.

### CODE REVIEW

**Comment by ackagel** [comment]
@cliu72 here's the updatable docker stuff.  There isn't a docker image uploaded for this one at the moment; I'll make one so you can just pull that instead of building this locally.  I'll let you know when that's up so you can confirm this updates ark code with that `-u` flag (I tested it on my end and it worked ok, but just a double check).

**Review by srivarra — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -23,20 +23,24 @@ RUN apt-get install -y gcc r-base
...
+RUN R -e "install.packages('arrow', repos = 'https://packagemanager.rstudio.com/all/__linux__/bullseye/latest')"
+
+#install flowsom requirements
+RUN apt-get install -y r-cran-igraph r-bioc-biocgenerics r-bioc-consensusclusterplus r-cran-dplyr r-cran-ggforce r-cran-ggplot2 r-cran-ggpubr r-cran-ggrepel r-cran-magrittr r-cran-pheatmap r-cran-rlang r-cran-rtsne r-cran-tidyr r-cran-xml r-cran-scattermore
+#install flowsom dependency requirements (eye-roll)
+RUN apt-get install -y r-cran-rcppparallel r-bioc-biobase r-cran-matrixstats r-cran-png r-cran-jpeg r-cran-interp r-cran-mass r-bioc-graph r-bioc-rbgl r-cran-scales r-cran-digest r-cran-bh r-cran-rcpparmadillo r-cran-jsonlite r-cran-base64enc r-cran-plyr r-bioc-zlibbioc r-cran-hexbin r-cran-gridextra r-cran-yaml r-bioc-rhdf5lib r-cran-corpcor r-cran-runit r-cran-tibble r-cran-xml2 r-cran-tweenr r-cran-gtable r-cran-polyclip r-cran-tidyselect r-cran-withr r-cran-lifecycle r-cran-rcppeigen
+
```
Comment (srivarra) [inline Dockerfile:37]: Oh my god.

Comment (ackagel) [reply Dockerfile:37]: yes i endured pain

**Review by cliu72 — COMMENTED** [review]
Just tested it and it works for me!

File: Dockerfile (hunk 1 above)
Comment (cliu72) [reply Dockerfile:37]: lol

**Comment by ngreenwald** [comment]
Will you also test modifying the `ark` codebase with jupyter launched, and see what %magic% command or whatever needs to be run within the notebook for the updated changes?

**Comment by ngreenwald** [comment]
Testing this is what I meant:
`cd /opt/ark-analysis && pip install `

**Comment by ackagel** [comment]
@cliu72 awesome! do you mind also running some R and/or pixel clustering stuff, just to make sure it works? I changed the way flowsom gets installed (for speed & less-"compiler error debugging" reasons), and just want to make sure everything works as intended.

Mostly made those changes because travis ran into a compiler error while trying to get this docker built that looked like it might bog down future docker builds on travis.  Bonus from that adventure, the docker compilation time is down from ~1hr to ~20mins

**Comment by cliu72** [comment]
> Will you also test modifying the `ark` codebase with jupyter launched, and see what %magic% command or whatever needs to be run within the notebook for the updated changes?

Yup tested this already. The specific commands are:
`cd /opt/ark-analysis`
`pip install .`

(double ampersand doesn't work)

Also @ackagel, I didn't actually need the `-u` tag to make this work. Without it, the above commands still work to make the code editable.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
I think the -u flag basically does that automatically. So for non technical users, they can just run that command if we've changed the code but not made a new release, and they'll get the updated version. For developers, they can run that command the first time to incorporate their changes, and then for subsequent changes with the docker still open, use that command within the notebook. Is that correct Adam?

How long did it take when you ran that in the notebook?

**Comment by cliu72** [comment]
Oh I see. So using the -u flag should pull the latest changes in ark, and running the cd/pip commands will pull any local changes the user makes. Is that right? If that's the case, I tested the second part of that (cd/pip part) and it works (not sure if you wanted me to test the first part of that too? not sure how to do that)

Running the pip install command in jupyter takes ~20 seconds.

Tested pixel clustering and it seems fine. Just a note - Alex created our own version of the FlowSOM package and edited it a bit. I don't think his changes are super obvious when testing, so I didn't confirm whether our version of FlowSOM is being used. Just wanted to confirm that with your changes, you're using our version of FlowSOM and not installing from Bioconductor.

**Comment by ackagel** [comment]
The -u flag doesn't pull changes; users still need to run `git pull`.  The -u flag just runs the cd/pip commands automatically, before opening up jupyter.

I can confirm that we're using our version of FLowSOM, not Bioconductor's.

@ngreenwald sounds like we're all good here, so I'll update the branch and merge it in post-standup

**Comment by ackagel** [comment]
@ngreenwald just need you to approve to unblock merging

**Review by ngreenwald — APPROVED** [review]
Looks good

**Pull request merged**